File: 11-container-with-most-water/11-container-with-most-water.py
class Solution:
    def maxArea(self, height: List[int]) -> int:
        start = 0
        end = len(height) - 1
        maxWater = 0
        
        while start < end: # 0(n)
            length = min(height[start], height[end]) # 0(2)
            width = end-start
            currentArea = length*width
            maxWater = max(maxWater, currentArea) # 0(2)
            if height[start] > height[end]:
                end-=1
            else:
                start+=1
        return maxWater
    # time complexity: 0(4n) --> 0(n)

File: 11-container-with-most-water/11-container-with-most-water.cpp
class Solution {
public:
    int maxArea(vector<int>& height) {
        int start = 0;
        int end = height.size() - 1;
        int ans = 0;
        int length = 0;
        int width = 0;
        int curr = 0;
            
        while(start < end){
            length = min(height[start], height[end]);
            width = end - start;
            curr = length * width;
            ans = max(ans, curr);
            if(height[start] > height[end]){
                end--;
            }
            else{
                start++;
            }
        }
        return ans;
    }
};